### PR TITLE
Live Templates

### LINKED ISSUE #136: Basic Surrounder support
Start with surrounding exprs with (). Perhaps also include a `loop{}` surrounder, too.

See [golang's `SurroundDescriptor`.](https://github.com/go-lang-plugin-org/go-lang-idea-plugin/blob/master/src/com/goide/editor/surround/GoSurroundDescriptor.java)

### PR DESCRIPTION
**Pull request by Turbo87**
This PR adds a few useful live templates. It also resolves #136 by using surrounder live templates for the the loops.

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/template/RustFileContextType.kt [hunk 1]
```diff
@@ -6,7 +6,7 @@ import com.intellij.psi.util.PsiUtilCore
 import org.rust.lang.RustLanguage
 
 
-class RustFileContextType : TemplateContextType("RUST_FILE", "Rust file") {
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/template/RustFileContextType.kt]: I'm not sure about this. `Rust` by itself is good, but we will also want `Rust expression`, `Rust module` soon (In fact, we can already implement them) soon.

Comment (Turbo87) [inline src/main/kotlin/org/rust/lang/template/RustFileContextType.kt]: You can still do that. We should have a single `RustFileContextType` class, and then multiple other context classes that use `RustFileContextType` as their parent context. (see third parameter of the `TemplateContextType` class)

**Comment by matklad** [comment]
Hm, templates are an interesting issue :) The Rust syntax is rather minimal, so we should try not to provide unnecessary templates. 

I am not sure about this bunch of `t?struct\d` templates. `struct0` seems superfluous: it saves only one semicolon. I'm not sure about `struct1` and `struct2` either. I wish that my structs had at most 2 fields =)

What do you think about providing a single

```
struct $NAME$ {
    $END$
}
```
c
template and also a `f`ield template

```
$PUB$ $NAME$: $TYPE$,
```

which is only available inside a struct?

**Comment by Turbo87** [comment]
sounds good to me

**Comment by Turbo87** [comment]
@matklad done

File: src/main/resources/org/rust/lang/liveTemplates/Rust.xml [hunk 2]
```diff
@@ -24,4 +24,142 @@
...
+
+    <template name="match" description="Match expression" toReformat="true" toShortenFQNames="true"
+              value="match $EXPRESSION$ {&#10;  $VAL1$ =&gt; $EXPRESSION1$,&#10;  _ =&gt; $EXPRESSION2$,&#10;}$END$">
+        <variable name="EXPRESSION" expression="" defaultValue="&quot;x&quot;" alwaysStopAt="true" />
+        <variable name="VAL1" expression="" defaultValue="&quot;true&quot;" alwaysStopAt="true" />
+        <variable name="EXPRESSION1" expression="" defaultValue="" alwaysStopAt="true" />
+        <variable name="EXPRESSION2" expression="" defaultValue="" alwaysStopAt="true" />
+        <context>
+            <option name="RUST_FILE" value="true" />
+        </context>
```
Comment (matklad) [inline src/main/resources/org/rust/lang/liveTemplates/Rust.xml]: I would suggest a more minimal

```
        match x {
            _ => $END$,
        }
```

The rationale is that the first match arm can be quite complicated, so it's better to exit template earlier. 

And once we have type inference and postfix templates, we would be able to do incredible stuff here!

**Comment by matklad** [comment]
@Turbo87 don't you want to implement Rust Struct template context for the `f` templeate?

**Comment by Turbo87** [comment]
@matklad too sleepy for that :wink: 

Truth is that I'm going to be on vacation for the next few weeks and I don't want to start any large tasks so close before leaving...

File: src/main/resources/org/rust/lang/liveTemplates/Rust.xml (hunk 2 above)
Comment (Turbo87) [inline src/main/resources/org/rust/lang/liveTemplates/Rust.xml]: done

**Comment by matklad** [comment]
@Turbo87 lets leave this until you come back then =)

I want to be every template absolutelly use

**Comment by matklad** [comment]
...full. It is way to easy to create bloated and useless templates. At least, this is my experience with most of snippet bundles out there :). I really hate when I have to delete something from an expanded snippet.

**Comment by Turbo87** [comment]
@matklad 4 weeks, long time... ;)

I've tried implementing a subcontext with

```kotlin
PsiUtilCore.getElementAtOffset(file, offset) is RustStructDeclArgs
```

and I've also tried looking at the parents, but it doesn't seem to work as expected. In a scenario like this:

```rust
struct Foo {
  <caret>
}
```

I get `PsiElement(<IDENTIFIER>)` for the element at the caret and `PsiFile` as it's direct parent. In other words: it doesn't seem to be possible to detect whether we're in a struct. Most likely this is because as soon as something is inserted the struct is no longer parsed correctly and there will no longer be a struct in the Psi tree that we could use for context detection.

File: src/main/kotlin/org/rust/lang/template/RustFileContextType.kt (hunk 1 above)
Comment (matklad) [inline src/main/kotlin/org/rust/lang/template/RustFileContextType.kt]: You are absolutely right here, thanks for clarification.

**Comment by matklad** [comment]
> 4 weeks, long time... ;)

:cry:

I will try to experiment with this snippets then. Let's close this PR for now then, OK?

**Comment by Turbo87** [comment]
feel free to merge, close or reuse them in any way you want :wink:

**Comment by matklad** [comment]
An update on snippets:

The Rust struct context and `f` template were merged into master. You were right about `f` not working due to broken parse tree. This is fixable with `pin` attributes.


I am still not convinced that we need struct/fn/enum templates. We *may* need a keyword completion for them, but is of low priority.

(for|if|while)[let] templates are nice, but they would be super cool if implemented as postfix templates. I don't know how the postfix templates work, so I can't estimate the ammount of work here.


test template seems interesting. May be we should have a `tmod` which expands to

``` 
#[test] 
mod test { 
    use super::*; 
} 
```

and creates a `TestMod` context. test then can work in that context, and possibly some other templates can too. I haven't written that much `Rust` tests :), so I don't really know.

**Comment by chriskrycho** [comment]
@matklad a `tmod` would be great, but should expand to this instead:

```rust
#[cfg(test)]
mod tests {
    use super::*;
}
```

That's the [current convention](https://doc.rust-lang.org/book/testing.html#the-tests-module), at least!

I'd also love a quick action to generate a test for a given function, so if you had `do_something()` it could generate `test_do_something()` for you, but that's definitely lower down the list of "just things that would be nice to have."

**Comment by matklad** [comment]
@chriskrycho added `tmod` https://github.com/intellij-rust/intellij-rust/commit/5c35e8f34863e31482cfcb9cf5acf9fd55ae1b11.

**Comment by chriskrycho** [comment]
👍🏼 Splendid!

**Comment by Turbo87** [comment]
@matklad did you see that this PR also solved #136 ?

**Comment by matklad** [comment]
@Turbo87 no :) 

Other plugins and IDEA seem to use `SurroundDescriptors` as suggested in #136. Probably we should use them too.

**Comment by Turbo87** [comment]
@matklad that looks like a quite complicated solution to something that can be solved with a simple live template IMHO

**Comment by matklad** [comment]
@Turbo87 ok, I don't see why we need a `SurroundDescriptor` here. I'd accept a PR with surround templates.

**Pull request closed without merging**